File: gems/decomplex/lib/decomplex/delta.rb
# frozen_string_literal: true

require "json"
require_relative "root_cause"

module Decomplex
  # Before/after delta. decomplex is otherwise a stateless snapshot;
  # this is the feedback loop CLAUDE.md's "fixing reduces complexity,
  # not adds" demands: did a commit collapse the targeted root-cause
  # cluster, or just displace the debt elsewhere?
  #
  # Identity is LINE-INSENSITIVE by construction: a finding's
  # fingerprint is its detector + RootCause entities + (file, method),
  # never its line. A finding that only moved lines is "persisted",
  # NOT resolved+new -- so editing unrelated code above a finding does
  # not register as churn. Pure set/multiset arithmetic over two
  # snapshots; no re-analysis.
  module Delta
    module_function

    SEP = "\t" # source slices are space-collapsed; never contain tabs

    # sections: Report#sections_data ; clusters: RootCause.cluster(...).
    # Returns a JSON-round-trippable Hash (string keys).
    def snapshot(sections, clusters)
      findings = Hash.new(0)
      details = Hash.new { |h, k| h[k] = [] }
      sections.each do |title, _tier, fs|
        next unless fs

        fs.each do |f|
          fp = fingerprint(title, f)
          findings[fp] += 1
          details[fp] << json_safe_finding(title, f)
        end
      end
      site = Hash.new(0)
      site_details = Hash.new { |h, k| h[k] = [] }
      sections.each do |title, _tier, fs|
        next unless fs

        fs.each do |f|
          detail = json_safe_finding(title, f)
          site_fingerprints(title, f).each do |sfp|
            site[sfp] += 1
            site_details[sfp] << detail
          end
        end
      end
      cl = {}
      clusters.each do |c|
        cl["#{c[:kind]}#{SEP}#{c[:token]}"] =
          { "n" => c[:n_detectors], "s" => c[:support],
            "fat" => !c[:fat_union].nil? && c[:fat_union] }
      end
      { "findings" => findings, "site_findings" => site,
        "details" => details, "site_details" => site_details, "clusters" => cl,
        "total" => findings.values.sum }
    end

    def fingerprint(detector, finding)
      ents = RootCause.entities(finding)
                      .map { |k, t| "#{k}:#{t}" }.sort.join(",")
      units = RootCause.finding_units(finding)
                       .map { |f, m| "#{f}##{m}" }.uniq.sort.join(",")
      [detector, ents, units].join(SEP)
    end

    # Per-LEAF identity: one fingerprint per (detector, entities,
    # single file#method). The aggregate `fingerprint` re-keys a whole
    # multi-site finding when ANY member changes, so its per-finding
    # add/resolve is unusable for "which specific sites did this commit
    # add/remove" on aggregate detectors. ADDITIVE: `fingerprint`,
    # `findings`, `total`, clusters are unchanged (still report-
    # reconciled); this only powers the extra `site_*` delta.
    def site_fingerprints(detector, finding)
      ents = RootCause.entities(finding)
                      .map { |k, t| "#{k}:#{t}" }.sort.join(",")
      RootCause.finding_units(finding)
               .map { |f, m| "#{f}##{m}" }.uniq
               .map { |u| [detector, ents, u].join(SEP) }
    end

    def json_safe_finding(detector, finding)
      json_safe_value(finding.merge(detector: detector))
    end

    def json_safe_value(value)
      case value
      when Hash
        value.to_h { |k, v| [k.to_s, json_safe_value(v)] }
      when Array
        value.map { |v| json_safe_value(v) }
      when Symbol
        value.to_s
      else
        value
      end
    end

    # base/head: snapshot Hashes (or JSON.parse of them). Returns the
    # structured delta.
    def diff(base, head)
      bf = base["findings"] || {}
      hf = head["findings"] || {}
      keys = (bf.keys | hf.keys)
      resolved = []
      added = []
      persisted = 0
      keys.each do |k|
        d = hf[k].to_i - bf[k].to_i
        if d.negative?  then resolved << [k, -d]
        elsif d.positive? then added << [k, d]
        else persisted += 1
        end
      end

      bc = base["clusters"] || {}
      hc = head["clusters"] || {}
      cl = { "grown" => [], "shrunk" => [], "new" => [], "resolved" => [] }
      (bc.keys | hc.keys).each do |k|
        b = bc[k]
        h = hc[k]
        if b.nil?  then cl["new"] << [k, h["s"]]
        elsif h.nil? then cl["resolved"] << [k, b["s"]]
        else
          ds = h["s"].to_i - b["s"].to_i
          (cl["grown"] << [k, b["s"], h["s"]] if ds.positive?)
          (cl["shrunk"] << [k, b["s"], h["s"]] if ds.negative?)
        end
      end

      # Site-level (per-leaf) delta -- the precise "which (file#method)
      # findings did this commit add/remove" that the aggregate
      # findings delta above cannot answer for multi-site detectors.
      bs = base["site_findings"] || {}
      hs = head["site_findings"] || {}
      s_added = []
      s_resolved = []
      s_persisted = 0
      (bs.keys | hs.keys).each do |k|
        d = hs[k].to_i - bs[k].to_i
        if d.negative?  then s_resolved << [k, -d]
        elsif d.positive? then s_added << [k, d]
        else s_persisted += 1
        end
      end

      {
        "resolved" => resolved.sort_by { |_, n| -n },
        "added" => added.sort_by { |_, n| -n },
        "persisted" => persisted,
        "site_resolved" => s_resolved.sort_by { |_, n| -n },
        "site_added" => s_added.sort_by { |_, n| -n },
        "site_persisted" => s_persisted,
        "totals" => { "base" => base["total"].to_i,
                      "head" => head["total"].to_i,
                      "delta" => head["total"].to_i - base["total"].to_i },
        "clusters" => cl
      }
    end

    def label(fp)
      d, ents, units = fp.split(SEP)
      "#{d} `#{ents.to_s.split(',').first || units.to_s.split(',').first}`"
    end

    def to_markdown(d)
      t = d["totals"]
      dir = t["delta"].zero? ? "no net change" :
            (t["delta"].negative? ? "REDUCED #{-t['delta']}" : "GREW #{t['delta']}")
      o = +"## Delta vs baseline\n\n"
      o << "_Line-insensitive. Net debt: #{dir} " \
           "(#{t['base']} -> #{t['head']}; #{d['persisted']} persisted, " \
           "#{d['added'].size} added, #{d['resolved'].size} resolved)._\n\n"
      unless d["clusters"]["shrunk"].empty? && d["clusters"]["resolved"].empty?
        o << "**Collapsed / shrunk root-cause clusters (the goal):**\n"
        d["clusters"]["resolved"].first(10).each do |k, s|
          o << "- `#{k.split(SEP).last}` -- #{s} findings -> GONE\n"
        end
        d["clusters"]["shrunk"].first(10).each do |k, b, h|
          o << "- `#{k.split(SEP).last}` -- #{b} -> #{h} findings\n"
        end
        o << "\n"
      end
      unless d["clusters"]["new"].empty? && d["clusters"]["grown"].empty?
        o << "**New / grown clusters (displaced debt -- watch):**\n"
        d["clusters"]["new"].first(10).each do |k, s|
          o << "- `#{k.split(SEP).last}` -- NEW, #{s} findings\n"
        end
        d["clusters"]["grown"].first(10).each do |k, b, h|
          o << "- `#{k.split(SEP).last}` -- #{b} -> #{h} findings\n"
        end
        o << "\n"
      end
      o << "**Top added findings:**\n"
      d["added"].first(15).each { |k, n| o << "- (+#{n}) #{label(k)}\n" }
      o << "- (none)\n" if d["added"].empty?

      sa = d["site_added"] || []
      sr = d["site_resolved"] || []
      unless sa.empty? && sr.empty?
        o << "\n**Site-level changes (precise -- exact file#method, " \
             "robust to membership churn):**\n"
        o << "- site-added #{sa.size}, site-resolved #{sr.size}, " \
             "site-persisted #{d['site_persisted']}\n"
        sa.first(25).each do |k, _|
          dd, _e, u = k.split(SEP)
          o << "  - + `#{u}` (#{dd})\n"
        end
        sr.first(25).each do |k, _|
          dd, _e, u = k.split(SEP)
          o << "  - - `#{u}` (#{dd})\n"
        end
      end
      o
    end
  end
end

File: gems/decomplex/src/decomplex/delta.rs
use crate::decomplex::report::ReportSection;
use crate::decomplex::root_cause::{self, Cluster};
use serde_json::{json, Map, Value};
use std::collections::BTreeMap;

const SEP: &str = "\t";

pub fn snapshot(sections: &[ReportSection], clusters: &[Cluster]) -> Value {
    let mut findings: BTreeMap<String, i64> = BTreeMap::new();
    let mut details: BTreeMap<String, Vec<Value>> = BTreeMap::new();
    for section in sections {
        for finding in &section.findings {
            let fp = fingerprint(&section.title, finding);
            *findings.entry(fp.clone()).or_insert(0) += 1;
            details
                .entry(fp)
                .or_default()
                .push(json_safe_finding(&section.title, finding));
        }
    }

    let mut site: BTreeMap<String, i64> = BTreeMap::new();
    let mut site_details: BTreeMap<String, Vec<Value>> = BTreeMap::new();
    for section in sections {
        for finding in &section.findings {
            let detail = json_safe_finding(&section.title, finding);
            for sfp in site_fingerprints(&section.title, finding) {
                *site.entry(sfp.clone()).or_insert(0) += 1;
                site_details.entry(sfp).or_default().push(detail.clone());
            }
        }
    }

    let mut cluster_values = Map::new();
    for cluster in clusters {
        cluster_values.insert(
            format!("{}{}{}", cluster.kind, SEP, cluster.token),
            json!({
                "n": cluster.n_detectors,
                "s": cluster.support,
                "fat": cluster.fat_union,
            }),
        );
    }
    let total = findings.values().sum::<i64>();
    json!({
        "findings": findings,
        "site_findings": site,
        "details": details,
        "site_details": site_details,
        "clusters": cluster_values,
        "total": total,
    })
}

pub fn fingerprint(detector: &str, finding: &Value) -> String {
    let mut entities = root_cause::entities(finding)
        .into_iter()
        .map(|entity| format!("{}:{}", entity.kind, entity.token))
        .collect::<Vec<_>>();
    entities.sort();
    let mut units = root_cause::finding_units(finding)
        .into_iter()
        .map(|(file, method)| format!("{file}#{method}"))
        .collect::<Vec<_>>();
    units.sort();
    units.dedup();
    [detector.to_string(), entities.join(","), units.join(",")].join(SEP)
}

pub fn site_fingerprints(detector: &str, finding: &Value) -> Vec<String> {
    let mut entities = root_cause::entities(finding)
        .into_iter()
        .map(|entity| format!("{}:{}", entity.kind, entity.token))
        .collect::<Vec<_>>();
    entities.sort();
    let entity_text = entities.join(",");
    let mut units = root_cause::finding_units(finding)
        .into_iter()
        .map(|(file, method)| format!("{file}#{method}"))
        .collect::<Vec<_>>();
    units.sort();
    units.dedup();
    units
        .into_iter()
        .map(|unit| [detector.to_string(), entity_text.clone(), unit].join(SEP))
        .collect()
}

pub fn json_safe_finding(detector: &str, finding: &Value) -> Value {
    let mut object = finding.as_object().cloned().unwrap_or_default();
    object.insert("detector".to_string(), Value::String(detector.to_string()));
    Value::Object(object)
}

pub fn diff(baseline: &Value, head: &Value) -> Value {
    let baseline = embedded_snapshot(baseline);
    let head = embedded_snapshot(head);
    let before = count_map(baseline, "site_findings");
    let after = count_map(head, "site_findings");
    let keys = before
        .keys()
        .chain(after.keys())
        .cloned()
        .collect::<std::collections::BTreeSet<_>>();

    let mut introduced = Vec::new();
    let mut resolved = Vec::new();
    let mut persisted = Vec::new();
    for fingerprint in keys {
        let old = before.get(&fingerprint).copied().unwrap_or(0);
        let new = after.get(&fingerprint).copied().unwrap_or(0);
        if new > old {
            introduced.push(delta_entry(head, &fingerprint, new - old));
        }
        if old > new {
            resolved.push(delta_entry(baseline, &fingerprint, old - new));
        }
        if old.min(new) > 0 {
            persisted.push(delta_entry(head, &fingerprint, old.min(new)));
        }
    }

    let introduced_total = delta_total(&introduced);
    let resolved_total = delta_total(&resolved);
    let persisted_total = delta_total(&persisted);
    json!({
        "introduced": { "total": introduced_total, "findings": introduced },
        "resolved": { "total": resolved_total, "findings": resolved },
        "persisted": { "total": persisted_total, "findings": persisted },
        "net": introduced_total - resolved_total,
    })
}

pub fn to_markdown(delta: &Value) -> String {
    let introduced = delta
        .pointer("/introduced/total")
        .and_then(Value::as_i64)
        .unwrap_or(0);
    let resolved = delta
        .pointer("/resolved/total")
        .and_then(Value::as_i64)
        .unwrap_or(0);
    let persisted = delta
        .pointer("/persisted/total")
        .and_then(Value::as_i64)
        .unwrap_or(0);
    let net = delta.get("net").and_then(Value::as_i64).unwrap_or(0);
    let mut out = format!(
        "# Decomplex Delta\n\n- Introduced: **{introduced}**\n- Resolved: **{resolved}**\n- Persisted: **{persisted}**\n- Net: **{net:+}**\n\n"
    );
    render_delta_section(
        &mut out,
        "Resolved structural findings",
        delta.pointer("/resolved/findings"),
    );
    render_delta_section(
        &mut out,
        "Introduced structural findings",
        delta.pointer("/introduced/findings"),
    );
    out
}

fn embedded_snapshot(value: &Value) -> &Value {
    value
        .pointer("/runs/0/properties/decomplex.snapshot")
        .unwrap_or(value)
}

fn count_map(snapshot: &Value, field: &str) -> BTreeMap<String, i64> {
    snapshot
        .get(field)
        .or_else(|| snapshot.get("findings"))
        .and_then(Value::as_object)
        .map(|values| {
            values
                .iter()
                .filter_map(|(key, value)| value.as_i64().map(|count| (key.clone(), count)))
                .collect()
        })
        .unwrap_or_default()
}

fn delta_entry(snapshot: &Value, fingerprint: &str, count: i64) -> Value {
    let details = snapshot
        .get("site_details")
        .or_else(|| snapshot.get("details"))
        .and_then(Value::as_object)
        .and_then(|values| values.get(fingerprint))
        .cloned()
        .unwrap_or_else(|| json!([]));
    json!({ "fingerprint": fingerprint, "count": count, "details": details })
}

fn delta_total(entries: &[Value]) -> i64 {
    entries
        .iter()
        .filter_map(|entry| entry.get("count").and_then(Value::as_i64))
        .sum()
}

fn render_delta_section(out: &mut String, title: &str, findings: Option<&Value>) {
    out.push_str(&format!("## {title}\n\n"));
    let findings = findings
        .and_then(Value::as_array)
        .cloned()
        .unwrap_or_default();
    if findings.is_empty() {
        out.push_str("None.\n\n");
        return;
    }
    for finding in findings {
        let count = finding.get("count").and_then(Value::as_i64).unwrap_or(0);
        let fingerprint = finding
            .get("fingerprint")
            .and_then(Value::as_str)
            .unwrap_or("");
        let detail = finding
            .get("details")
            .and_then(Value::as_array)
            .and_then(|values| values.first())
            .and_then(Value::as_object)
            .and_then(|object| object.get("at").or_else(|| object.get("detail")))
            .and_then(Value::as_str)
            .unwrap_or(fingerprint);
        out.push_str(&format!("- `{detail}` — {count}\n"));
    }
    out.push('\n');
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn delta_credits_resolved_and_consolidated_findings() {
        let baseline = json!({
            "site_findings": { "tuple": 1, "grammar": 2 },
            "site_details": {
                "tuple": [{ "detail": "semantic tuple pressure" }],
                "grammar": [{ "detail": "duplicated grammar branch" }]
            }
        });
        let head = json!({
            "site_findings": { "grammar": 1, "record": 1 },
            "site_details": {
                "grammar": [{ "detail": "duplicated grammar branch" }],
                "record": [{ "detail": "new record issue" }]
            }
        });

        let result = diff(&baseline, &head);
        assert_eq!(
            result.pointer("/resolved/total").and_then(Value::as_i64),
            Some(2)
        );
        assert_eq!(
            result.pointer("/introduced/total").and_then(Value::as_i64),
            Some(1)
        );
        assert_eq!(
            result.pointer("/persisted/total").and_then(Value::as_i64),
            Some(1)
        );
        assert_eq!(result.get("net").and_then(Value::as_i64), Some(-1));
        let markdown = to_markdown(&result);
        assert!(markdown.contains("semantic tuple pressure"));
        assert!(markdown.contains("duplicated grammar branch"));
    }

    #[test]
    fn delta_reads_snapshots_embedded_in_sarif() {
        let sarif = json!({ "runs": [{ "properties": { "decomplex.snapshot": {
            "site_findings": { "old": 1 }
        } } }] });
        let result = diff(&sarif, &json!({ "site_findings": {} }));
        assert_eq!(
            result.pointer("/resolved/total").and_then(Value::as_i64),
            Some(1)
        );
    }
}
